kiwi/lib/search.py

```diff
@@ -0,0 +1,655 @@
+        #   added by `kiwi.cli.arguments_to_configuration` (and we remove them
+        #   so Pydantic won't throw an error.)
+        del base_dict['verbose'], base_dict['quiet']
+        base_config = train.Configuration(**base_dict)
+    else:
+        base_config = config.base_config
+    base_config_dict = base_config.dict()
+
+    # Perform some checks of the training config
+    use_mlflow = True if base_config_dict['run'].get('use_mlflow') else False
+    if not use_mlflow:
+        logger.warning(
+            'Using MLflow is recommend; set `run.use_mlflow=true` in the base config'
+        )
+    # The main metric should be explicitly set
+    if base_config_dict['trainer']['main_metric'] is None:
+        raise ValueError(
+            'The metric should be explicitly set in `trainer.main_metric` '
+            'in the training config (`base_config`).'
+        )
+    # The main metric(s) should be inside a list
+    if not isinstance(base_config_dict['trainer']['main_metric'], list):
+        base_config_dict['trainer']['main_metric'] = [
+            base_config_dict['trainer']['main_metric']
+        ]
+
+    # Use the early stopping logic of the Kiwi trainer
+    base_config_dict['trainer']['checkpoint'][
+        'early_stop_patience'
+    ] = config.options.patience
+    base_config_dict['trainer']['checkpoint'][
+        'validation_steps'
+    ] = config.options.validation_steps
+
+    # Load or initialize a study
+    if config.load_study:
+        logger.info(f'Loading study to resume from: {config.load_study}')
+        study = joblib.load(config.load_study)
+    else:
+        if config.options.search_method == 'random':
+            logger.info('Exploring parameters with random sampler')
+            sampler = optuna.samplers.RandomSampler(seed=config.seed)
+        else:
+            multivariate = (config.options.search_method == 'multivariate_tpe',)
+            logger.info(
+                'Exploring parameters with '
+                f'{"multivariate " if multivariate else ""}TPE sampler'
+            )
+            sampler = optuna.samplers.TPESampler(
+                seed=config.seed, multivariate=multivariate,
+            )
+        logger.info('Initializing study...')
+        pruner = optuna.pruners.MedianPruner()
+        study = optuna.create_study(
+            study_name=config.search_name,
+            direction='maximize',
+            pruner=pruner,
+            sampler=sampler,
+        )
+
+    # Optimize the study
+    optimize_kwargs = {}
+    if use_mlflow:
+        optimize_kwargs['callbacks'] = [optuna.integration.MLflowCallback()]
+    try:
+        logger.info('Optimizing study...')
+        objective = Objective(config=config, base_config_dict=base_config_dict)
+        study.optimize(
+            objective, n_trials=config.num_trials, **optimize_kwargs,
+        )
+    except KeyboardInterrupt:
+        logger.info('Early stopping search caused by ctrl-C')
+    except Exception as e:
```

Comment:
Terrible practice, don't catch generic Exceptions unless there's a very good reason for this. I guess you want to still save the study even if an exception stops it half-way.

not a terrible use case, but I'd prefer real Exceptions that might actual show up.

Reply:
Yes. But. For this situation, we really want to catch *everything*. Because image if that study does not get saved. That would be such a waste.

Reply:
Understandable, saving the study is important.